File: Javascript/539.minimum_time_difference [2].js
/**
 * @param {string[]} timePoints
 * @return {number}
 */
var findMinDifference = function(timePoints) {
    const s = new Set(timePoints)
    // presence of duplicate items mean min-diff is 0
    if(s.size != timePoints.length)
        return 0
    
    const times = []
    for(const hhmm of s){
        const hh = hhmm.substring(0, 2)
        const mm = hhmm.substring(3)
        const time = 60*parseInt(hh) + parseInt(mm)

        times.push(time)
    }

    times.sort((a, b) => a - b)

    const m = times.length
    let minDiff = 24*60
    for(let i=0; i<m; i++){
        let diff = Math.abs(times[(i+1) % m] - times[i])
        if(diff > 12*60)
            diff = 24*60 - diff

        minDiff = Math.min(minDiff, diff)
    }

    return minDiff
};

File: CPP [0501-1000]/539.minimum-time-difference [2].cpp
class Solution {
public:
    // T.C.=O(n + m*log(m)), S.C.=O(m)
    // n: size of timePoints[], m: min(n, 24*60)
    int findMinDifference(vector<string>& timePoints) {
        unordered_set s(timePoints.begin(), timePoints.end());
        // presence of duplicate items mean min-diff is 0
        if(s.size() != timePoints.size())
            return 0;

        vector<int> times;
        for(string hhmm: s)
        {
            string hh = hhmm.substr(0, 2);
            string mm = hhmm.substr(3);
            int time = 60*stoi(hh) + stoi(mm);

            times.push_back(time);
        }

        sort(times.begin(), times.end());

        int m=times.size();
        int minDiff = 24*60;
        for(int i=0; i<m; i++)
        {
            int diff = abs(times[(i+1) % m] - times[i]);
            if(diff > 12*60)
                diff = 24*60 - diff;

            minDiff = min(minDiff, diff);
        }
        
        return minDiff;
    }
};